rudder-core/src/main/scala/com/normation/rudder/services/reports/NodeChangesService.scala

```diff
@@ -33,102 +33,268 @@
+  /**
+   * For the given service, what are the intervals currently valid?
+   * Optionnaly give a starting date, or else the service is in charge
+   * to provide a default value.
+   */
+  def getCurrentValidIntervals(since: Option[DateTime]): Seq[Interval]
 
-  override def getChanges(notBefore: Option[DateTime]): Box[Seq[ResultRepairedReport]] = {
-    reportsRepository.getChangeReports(notBefore.getOrElse(DateTime.now.minusDays(changesMaxAge)))
-  }
+  /*
+   * The two following methods are pure algorithm methods and we want that all
+   * implementation of getChangesByInterval use them.
+   */
 
   /**
-   * Get all changes until a date specified as parameter and regroup them by interval of hours
+   * Get interval
    */
-  override def getChangesByInterval(intervalLength : Int = 6, notBefore: Option[DateTime]  = None)={
+  def filterGroupByInterval(intervals: Seq[Interval], changes: Seq[ResultRepairedReport]): ChangesByRule = {
+    changes.groupBy(_.ruleId).mapValues { ch =>
+      ( for {
+         interval <- intervals
+      } yield {
+        (interval,ch.filter(interval contains _.executionTimestamp))
+      } ).toMap
+    }
+  }
 
-    // Limit of changes.
-    val limit = notBefore.getOrElse(DateTime.now.minusDays(changesMaxAge))
-    val now = DateTime.now
 
+  def getInterval(since: DateTime, to: DateTime): Seq[Interval] = {
     // We want our first interval limit to be 00:00 | 06:00 | 12:00 o| 18:00
+
+    val intervals = Seq(
+
+    )
+
     val firstInterval = {
       val firstHour =
-        if ( now isBefore now.withHourOfDay(6) ) {
-          now.withHourOfDay(0)
+        if ( to isBefore to.withHourOfDay(6) ) {
+          to.withTimeAtStartOfDay()
         } else {
-          if (now isBefore now.withHourOfDay(12)) {
-            now.withHourOfDay(6)
+          if (to isBefore to.withHourOfDay(12)) {
+            to.withTimeAtStartOfDay().withHourOfDay(6)
           } else {
-            if (now isBefore now.withHourOfDay(18)) {
-              now.withHourOfDay(12)
+            if (to isBefore to.withHourOfDay(18)) {
+              to.withTimeAtStartOfDay().withHourOfDay(12)
             } else {
-              now.withHourOfDay(18)
+              to.withTimeAtStartOfDay().withHourOfDay(18)
         } } }
       // Fix minutes to 0
-      firstHour.withMinuteOfHour(0)
+      new Interval(firstHour, firstHour.plusHours(6))
     }
 
     // Compute intervals of changes recursively from the last interval lower
-    def computeIntervals(previousLowerBound : DateTime)  : List[Interval] = {
-      if ( previousLowerBound isBefore limit) {
+    def computeIntervals(previousLowerBound: DateTime): List[Interval] = {
+      if ( previousLowerBound isBefore since) {
         // Our last bound is before the limit, do not create a interval
         Nil
       } else {
-        val nextLowerBound = previousLowerBound.minusHours(intervalLength)
+        val nextLowerBound = previousLowerBound.minusHours(6)
```

Comment:
It was already, but saying it wasn't. All computation were assuming 6h, or even more preciselly the 4 intervals [0h-6h[, [6h-12h[, [12h-18h[, [18h-24h[

So at least it's explicit, and if one day we want to change that, we DO know what are the hypothesis.